syft/create_sbom_config.go

```diff
@@ -0,0 +1,312 @@
+// is not visible to the syft API, but would be useful to see in the SBOM output.
+func (c CreateSBOMConfig) WithTool(name, version string, cfg ...any) CreateSBOMConfig {
+	c.ToolName = name
+	c.ToolVersion = version
+	c.ToolConfiguration = cfg
+	return c
+}
+
+// WithParallelism allows for setting the number of concurrent cataloging tasks that can be performed at once
+func (c CreateSBOMConfig) WithParallelism(p int) CreateSBOMConfig {
+	if p < 1 {
+		// TODO: warn?
+		p = 1
+	}
+	c.Parallelism = p
+	return c
+}
+
+// WithSearchConfig allows for setting the specific search configuration for cataloging.
+func (c CreateSBOMConfig) WithSearchConfig(cfg cataloging.SearchConfig) CreateSBOMConfig {
+	c.Search = cfg
+	return c
+}
+
+// WithRelationshipsConfig allows for defining the specific relationships that should be captured during cataloging.
+func (c CreateSBOMConfig) WithRelationshipsConfig(cfg cataloging.RelationshipsConfig) CreateSBOMConfig {
+	c.Relationships = cfg
+	return c
+}
+
+// WithDataGenerationConfig allows for defining what data elements that cannot be discovered from the underlying
+// target being scanned that should be generated after package creation.
+func (c CreateSBOMConfig) WithDataGenerationConfig(cfg cataloging.DataGenerationConfig) CreateSBOMConfig {
+	c.DataGeneration = cfg
+	return c
+}
+
+// WithPackagesConfig allows for defining any specific behavior for syft-implemented catalogers.
+func (c CreateSBOMConfig) WithPackagesConfig(cfg pkgcataloging.Config) CreateSBOMConfig {
+	c.Packages = cfg
+	return c
+}
+
+// WithFilesConfig allows for defining file-based cataloging parameters.
+func (c CreateSBOMConfig) WithFilesConfig(cfg filecataloging.Config) CreateSBOMConfig {
+	c.Files = cfg
+	return c
+}
+
+// WithoutFiles allows for disabling file cataloging altogether.
+func (c CreateSBOMConfig) WithoutFiles() CreateSBOMConfig {
+	c.Files = filecataloging.Config{
+		Selection: file.NoFilesSelection,
+		Hashers:   nil,
+	}
+	return c
+}
+
+// WithCatalogerSelection allows for adding to, removing from, or sub-selecting the final set of catalogers by name or tag.
+func (c CreateSBOMConfig) WithCatalogerSelection(selection pkgcataloging.SelectionRequest) CreateSBOMConfig {
+	c.CatalogerSelection = selection
+	return c
+}
+
+// WithoutCatalogers removes all catalogers from the final set of catalogers. This is useful if you want to only use
+// user-provided catalogers (without the default syft-provided catalogers).
+func (c CreateSBOMConfig) WithoutCatalogers() CreateSBOMConfig {
+	c.packageTaskFactories = nil
+	c.persistentPackageTaskFactories = nil
+	return c
+}
+
+// WithCatalogers allows for adding user-provided catalogers to the final set of catalogers that will always be run
+// regardless of the source type or any cataloger selections provided.
+func (c CreateSBOMConfig) WithCatalogers(catalogers ...pkg.Cataloger) CreateSBOMConfig {
```

Comment:
from @willmurphyscode : consider removing this and adding to the signature for `CreateSBOM()`. This would make the nuance a little more clear. This also implies `WithCataloger ` would not make a distinction between 0-tags (tags are required?).

Reply:
consider this as well:
```
WithCataloger(cat pkg.Cataloger, forceEnabled bool, tags ...string)
```

this is harder to misuse given that the persistence vs selectability of the cataloger is explicitly given.

Reply:
another approach, we could make this a rich struct that has all of these options.
```golang
type CatalogerConfig struct {
  Cataloger pkg.Cataloger
  ForceEnabled bool
  Tags []string
}

func WithCatalogers(CatalogerConfig ...) CreateSBOMConfig {

}

// maybe not?
func WithSelectableCataloger(pkg.Cataloger, tags []string) CreateSBOMConfig {

}

func WithAlwaysEnabledCatalogers(pkg.Cataloger ...) CreateSBOMConfig {

}
```

Reply:
I went with this last option